queries/purescript/highlights.scm

```diff
@@ -0,0 +1,149 @@
+; ----------------------------------------------------------------------------
+; Literals and comments
+
+ (integer) @number
+ (exp_negation) @number
+ (exp_literal (number)) @float
+ (char) @character
+ [
+   (string)
+   (triple_quote_string)
+ ] @string
+
+ (comment) @comment
+
+
+; ----------------------------------------------------------------------------
+; Punctuation
+
+ [
+   "("
+   ")"
+   "{"
+   "}"
+   "["
+   "]"
+ ] @punctuation.bracket
+
+ [
+ (comma)
+ ";"
+ ] @punctuation.delimiter
+
+
+; ----------------------------------------------------------------------------
+; Keywords, operators, includes
+
+ ; This needs to come before the other "else" in
+ ; order to be highlighted correctly
+ (class_instance "else" @keyword)
+
+ [
+   "if"
+   "then"
+   "else"
+   "case"
+   "of"
+ ] @conditional
+
+ [
+   "import"
+   "module"
+ ] @include
+
+ [
+   (operator)
+   (constructor_operator)
+   (type_operator)
+   (qualified_module)  ; grabs the `.` (dot), ex: import System.IO
+   (all_names)
+
+   ; `_` wildcards in if-then-else and case-of expressions,
+   ; as well as record updates and operator sections
+   (wildcard)
+   "="
+   "|"
+   "::"
+   "∷"
+   "=>"
+   "⇒"
+   "<="
+   "⇐"
+   "->"
+   "→"
+   "<-"
+   "←"
+   "\\"
+   "`"
+   "@"
+ ] @operator
+
+ (qualified_module (module) @constructor)
+ (module) @namespace
+ (qualified_type (module) @namespace)
+ (qualified_variable (module) @namespace)
+ (import (module) @namespace)
+
+ [
+   (where)
+   "let"
+   "in"
+   "class"
+   "instance"
+   "derive"
+   "foreign"
+   "data"
+   "newtype"
+   "type"
+   "as"
+   "hiding"
+   "do"
+   "ado"
+   "forall"
+   "∀"
+   "infix"
+   "infixl"
+   "infixr"
+ ] @keyword
```

Comment:
Some of these might be `@type.qualifier` (i.e. modifiers). "forall" could be `@repeat`.

Reply:
Sorry, it's not obvious to me which ones you have in mind for `@type.qualifier`?

"forall" is "for all" precisely as in logic (universal quantification), not the for-loop.

Reply:
I was thinking the infix ones and maybe also "foreign".

Reply:
The infix ones are not related to types, they declare an operator (syntax), like this:
```
infixl 6 add as + -- now you can use '+'
```

`foreign` also does not modify the type of what it's directed at, it's a keyword (or more specifically, part of it) for using foreign function interface, so the type of a value does not depend on whether it's coming from `foreign` or not. I also don't think it fits well with `@include` because the actual inter-module imports in PS are strictly separated from the values declaration and can only happen at the top of a module, whereas foreign definitions can occur anywhere. also foreign definitions must come from a file with the same name but different extension, so it's more of a "same place" thing rather than a proper "import" which deals with actual outside files.

Reply:
The `@type.qualifier` is an unfortunate name. Think of them as modifiers.

Reply:
I see, but in `CONTRIBUTING.md` it's listed in the `Types` section, while `infix[lr]` and `foreign import` don't in any way act on types and are semantically closer to JS' `function` or Python's `def` keyword. They don't _define_ a function however because the definition happens elsewhere, so I think `@keyword` is the most sensible and least ambiguous choice here.

Reply:
Kotlin uses `@type.qualifier` for `infix`.

Reply:
Haskell uses `@keyword`, as well as Github's and VS Code's Atom-based grammars and [TryPureScript](https://try.purescript.org) highlights. I tried running it with `@type.qualifier` and I noticed that even among the most popular and "recommended" (listed in the wiki) themes, only a few of them define a separate color for `@type.qualifier` (everforest, nord). The rest either just don't define it which makes it fallback to `@type`, which in turn looks really bad because it's a keyword blending into non-keyword "area", or define it the same as `@keyword` (kanagawa, catppuccin, sonokai). In fact among 11 themes only 2 make `@type.qualifier` styled differently from `@type`/`@keyword`, so effectively it's a keyword 40% of the time and type (wrong) 40% of the time, which IMO is not worth it.

Reply:
Isn't that an issue w/ the themes and not us then? It's nice to be more accurate here, and while I don't know purescript, some of these keywords do look like qualifiers/modifiers

Reply:
Isn't that an issue w/ the themes and not us then? It's nice to be more accurate here, and while I don't know purescript, some of these keywords do look like qualifiers/modifiers

Reply:
It's also an issue with us having an inaccurate name but there's not much of a reason to change it before 1.0.

Reply:
I think accuracy is nice but I see neither an accuracy nor a looks gain here, and I described above what could be lost. Of course it would be impossible by definition for different languages to have the same mapping to Neovim's capture groups, but it seems more natural to me in the face of ambiguity to opt for the choice that is more generic and also results in better UX.